<!doctype html>
<html lang="en">
  <nav class="navbar navbar-expand-lg navbar-dark bg-dark sticky-top">
    <a class="navbar-brand" href="#">
      <img src="Hand-Print-and-Paw-Print-Decal.png" width="30" height="30" alt="">
      FREYA
    </a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavDropdown" aria-controls="navbarNavDropdown" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>
    <div class="collapse navbar-collapse" id="navbarNavDropdown">
      <ul class="navbar-nav">
        <li class="nav-item">
          <a class="nav-link" href="#DataPrep">DataPrep</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="#DataAnalysis">DataAnalysis</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/flatironinstitute/FREYA">Install FREYA</a>
        </li>
      </ul>
    </div>
  </nav>

  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>FREYA</title>

    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.1/css/bootstrap.min.css" integrity="sha384-WskhaSGFgHYWDcbwN70/dfYBj47jz9qbsMId/iRN3ewGhXQFZCSftd1LZCfmhktB" crossorigin="anonymous">

    <!-- Custom styles for this template -->
    <link href="product.css" rel="stylesheet">
  </head>

  <body class="background-color">
  <div class="background-gradient">
<!--      <nav class="site-header sticky-top py-1 logo">
        <a class="py-2" href="#"><img width="40" src="Hand-Print-and-Paw-Print-Decal.png" >FREYA</a>
        <div class="container d-flex flex-column flex-md-row justify-content-between nav" style="padding-top: 15px" >
          <a class="py-2 d-none d-md-inline-block" href="#DataPrep">DataPrep</a>
          <a class="py-2 d-none d-md-inline-block" href="#DataAnalysis">DataAnalysis</a>
          <a class="py-2 d-none d-md-inline-block" href="https://github.com/flatironinstitute/FREYA#readme">Install FREYA</a>
        </div>
      </nav>
--!>

      <div class="p-3 p-md-5 text-md-center"><img src="Hand-Print-and-Paw-Print-Decal.png" alt="" width="301" height="301" class="icon"/>
        <div class="col-md-5 p-lg-5 mx-auto my-5 offset-xl-3 col-xl-7 title" align="left">
          <h1 class="display-1 font-weight-bold">FREYA</h1>
          <p class="lead font-weight-normal"><b>FR</b>amework for <b>E</b>xpression anal<b>Y</b>sis <b>A</b>cross species</p>
          <a class="btn-outline-secondary btn-lg btn link-color" href="https://github.com/flatironinstitute/FREYA#readme"><img class="mr-2" alt="github" src="GitHub-Mark/PNG/GitHub-Mark-32px.png">Install FREYA</a>
        </div>
    </div>

  </div>
    <div class="d-md-flex flex-md-equal w-100 my-md-3 pl-md-3">
      <div class="mr-md-3 pt-3 px-3 pt-md-5 px-md-5 text-center overflow-hidden">
        <div class="my-3 p-3">
          <h2 class="display-5">FREYA Framework</h2>
          <p class="lead">A computational pipeline for cross-species cancer analyses</p>
          <!-- TODO image map for the links in the image -->
          <div class="media"><img class="mr-3 img-fluid" src="Figure1_Pipeline.png" alt="FREYA Framework"></div>
          </div>
        <p class="lead font-weight-normal" style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%"> The FREYA analytic framework calculates genetic diversity within a cancer cohort, extracts progression-related patterns of expression, calls functional variants, identifies intrinsic human molecular tumor subtypes, and compares them to human breast cancer biology.
          It helps streamline future canine mammary tumor (or other tissue/species) analyses and provides a comprehensive suite of tools that encompass conventional human analyses and new dog-centric approaches, seamlessly integrating the two. </p>
        <br>
        <p class="lead font-weight-normal" style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%"> FREYA is split into two major components: data preparation <b>(DataPrep)</b> and analysis <b>(DataAnalysis)</b>.
          DataPrep prepares sequencing data for analysis. 
          DataAnalysis runs statistical analyses from the manuscript.
          Both DataPrep and DataAnalysis are designed to run with user-provided data. 
          The pipeline requires that each patient (e.g. each dog) has at least 1 sample of each histology: normal, benign, and malignant. </p>
        </br>
        <a class="btn-outline-secondary btn-lg btn btn-dark link-color" href="https://github.com/flatironinstitute/FREYA#readme"><img class="mr-3" src="GitHub-Mark/PNG/GitHub-Mark-32px.png">Install FREYA</a> <br>
        <br>
      </div>
	  
    </div>
      
<!-- INTRODUCTION -->

<!-- DATAPREP -->
    <div class="d-md-flex flex-md-equal w-100 my-md-3 pl-md-3" id="DataPrep">
      <div class="mr-md-3 pt-3 px-3 pt-md-5 px-md-5 text-center overflow-hidden">
        <div class="my-3 p-3">
          <h1 class="display-5">Preparing the Data </h1>
          <p class="lead">( DataPrep )</p>
        </div>
        <p class="lead font-weight-normal text-left" style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
          DataPrep is the first sub-pipeline of FREYA. 
          It will prepare processed expression and mutation call data, using the raw data downloaded from your sequencer.
          DataPrep runtime is highly dependent on the machine used. 
          For example, the dataset used in the CMT manuscript takes approximately 1 month to process using 1 cpu but can be run in less than 24 hours using the DisBatch setup.
        </p>
        <p class="lead font-weight-normal text-left" style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
          This pipeline has several dependencies, which are, in effect, described by the Dockerfile. 
          The central script is "<kbd>cmwf_csv.sh</kbd>."
          Please see our GitHub documentation for more information. 
          Mutations are called using <a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=3891">GATK Best Practices workflow for SNP and indel calling on RNAseq data</a>.
        </p>
<!-- TODO: Add links for install for all the dependencies -->
        <br>
        <br>
        <h3> Workflow Steps and Environment Requirements </h3>
        <br>
        <h4 class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">Required Input: </h4>
        <ul class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
          <li>High throughput sequencing (<kbd>.fastq</kbd>) files for each sample</li>
          <li>Comma-separated phenotype file containing sample information: histology, sample ID, and patient. Multiple samples per patient are required. </li>
          <li>Reference genome files (<kbd>.gtf</kbd>, <kbd>.gff</kbd>,  etc)</li>
        </ul>
        <h4 class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">Dependencies:</h4>
        <ul class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
          <li>Python, Java, fastQC, DEXSeq, Samtools</li>
          <li>JAR files: GATK, Picard, SnpEff</li>
        </ul>
        <h4 class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%" >Steps:</h4>
        <ul class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
          <li>Generate RNA-Seq counts
            <ul class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
              <li>Align to the reference genome </li>
              <li>Generate quality reports</li>
              <li>Construct exon counts</li>
              <li>Mark duplicate reads</li>
              <li>Reassign mapping quality scores</li>
            </ul>
          </li>
          <li>Call variants, annotate to genes, create genes-by-samples mutation files
            <ul class='text-left'  style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
              <li>Call SNPs and indels</li>
              <li>Variant filtration</li>
              <li>Add gene name annotations</li>
              <li>Select final mutation calls</li>
            </ul>
          </li>
        </ul>
        <a class="btn btn-outline-secondary btn-lg link-color" href="https://github.com/flatironinstitute/FREYA#readme"><img class="mr-3" alt="github" src="GitHub-Mark/PNG/GitHub-Mark-32px.png">Install FREYA</a>
        <br>
        <br>
        <br>
      </div>
      <br>
    </div>

<!-- DATAANALYSIS -->
    <div class="d-md-flex flex-md-equal w-100 my-md-3 pl-md-3" id="DataAnalysis">
      <div class="mr-md-3 pt-3 px-3 pt-md-5 px-md-5 text-center overflow-hidden">
        <div class="my-3 p-3">
          <h1 class="display-5">Analyzing the Data </h1>
          <p class="lead">( DataAnalysis )</p>
        </div>
        <div style="width: 90%; border-radius: 21px 21px 21px 21px; padding-left: 10%">
          <p class="lead font-weight-normal text-left">
           This second sub-pipeline depends on processed genomic and phenotype data (created either by yourself or by using FREYA DataPrep).
           If you used DataPrep, be sure to first run the prep_data.R script to convert the gene names etc.
           If you are providing pre-processed data you may be able to skip this step (see below).
           </p><p class="lead font-weight-normal text-left">
           FREYA DataAnalysis can be run from your browser using either Docker and the Jupiter notebooks, or on the command line by using the provided Makefile.
           Update the variables in the Jupyter notebook, following the order in Index.ipynb, or in the Makefile, to reflect your data files.
           We provide some example simulated data in the synthetic_data folder.
           If you want to view or run the pipeline without using your own data, you can do so by clicking <a href="https://mybinder.org/v2/gh/flatironinstitute/FREYA/master">Launch Binder</a> (also in the Github repository).
          </p>
         <br>
         <br>
         <h3> Workflow Steps and Environment Requirements </h3>
         <br>
         <h4 class='text-left'>Required Input: </h4>
         <ul class='text-left'>
         <li>Expression matrix file</li>
         <li>Comma-separated phenotype file containing sample information: histology, sample ID, patient ID.</li>
         </ul>
         <h4 class='text-left'>Dependencies:</h4>
         <ul class='text-left'>
         <li>R</li>
         <li>Docker (if using Docker version of the pipeline)</li>
         </ul>
<!-- TODO: Add links for install for all the dependencies -->
         <h4 class='text-left'>Steps:</h4>
         <ul class='text-left'>
          <li>Prep the data - This preprocessing script creates the input files for the rest of the analysis scripts (except CMT_PEPs.csv).  </li>
          <li>Generate PEPS, run histology simulations - Generates new PEP lists based on the user data. </li>
          <li>Gene Ontology enrichment - runs GO enrichment using topGO. This requires a file with PEP enrichment scores. </li> 
          <li>PAGE enrichment - Calculates PAGE enrichment for a number of comparisons in the human data, using PEP lists from the canine data.  </li>
          <li>Differential Expression Analysis - Compares differential expression in human tumor/normal samples against the PEP genes from dog. </li>
          <li>PAM50 Subtype Analysis - Predicts PAM50 subtypes for the user data, based on TCGA BRCA RNA-Seq data. </li>
          <li>Mutation Analysis - Runs all mutations analysis on the user data. </li>
         </ul>
        <br>
         <h3 class='text-left'> Running the DataAnalysis Workflow </h3>
         <h4 class='text-left'>Running Using the Docker image</h4>
         <h5 class='text-left'>Setting up the Docker image</h5>
          <div class="media"><img class="mx-auto" style="width: 90%; margin-top: -30px; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box;" src="FREYA_SetUpDataAnalysis_1.png" alt="Setting up the Docker image"></div>
          <br>
         <h5 class='text-left' style="margin-top: 30px; ">How it appears in your browser</h5>
          <div class="media"><img class="mx-auto" style="width: 90%; margin-top: 18px;" src="FREYA_AppearanceDataAnalysis copy.png" alt="Setting up the Docker image"></div>
          <br>
         <h4 class='text-left'>Running Using Make</h4>
         <ol class="text-left">
         <li>Set up the repository and a folder containing your sample data. </li>
         <li>Update the variables at the top of the Makefile to point to your data, then run "<kbd>make clean all</kbd>".</li>
         <li>If you do not want to regenerate the peps and run data preprocessing, run "<kbd>make report</kbd>". </li>
         <li>The default output directory is 'results'</li>
         </ol>
        <a class="btn btn-outline-secondary btn-lg link-color" href="https://github.com/flatironinstitute/FREYA#readme"><img class="mr-3" alt="github" src="GitHub-Mark/PNG/GitHub-Mark-32px.png">Install FREYA</a>
        <br>
        <br>
        <br>
        </div>
        <br>
      </div>
    </div>


<!-- footer -->
    <footer class="container py-5">
      <div class="text-center"><a><img width="32" alt="Flatiron Institute" src="FL_Icon.png"></a></div>
    </footer>


    <!-- Bootstrap core JavaScript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.1/js/bootstrap.min.js" integrity="sha384-smHYKdLADwkXOn1EmN1qk/HfnUcbVRZyYmZ4qpPea6sjB/pTJ0euyQp0Mk8ck+5T" crossorigin="anonymous"></script>

  </body>
</html>
