{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PAGE analysis\n",
    "\n",
    "This is a Jupyter notebook.\n",
    "\n",
    "To run all cells in the notebook use `Cell --> Run All`.\n",
    "\n",
    "To run cells one at a time click into the first code cell and key `Shift-Enter` in each cell in sequence.\n",
    "\n",
    "More information on Jupyter notebooks can be found\n",
    "[here](http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Running%20Code.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set the plot window sizes within the notebook (in inches)\n",
    "options(repr.plot.width=6, repr.plot.height=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Please comment or uncomment the following declarations as appropriate for your run of this notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "opt <- list()\n",
    "\n",
    "# Set defaults for optional parameters\n",
    "opt$PEP = 'CMT_peps.csv'       ## PEP lists file (generated by the PEP creation script/notebook)\n",
    "opt$outdir = './results'        ## Output directory\n",
    "opt$datadir = './data'         ## Directory containing data provided with FREYA\n",
    "opt$workingdir = './user_data' ## Directory containing user data\n",
    "\n",
    "opt$outdir <- paste(opt$workingdir,opt$outdir, sep='/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## If the output directory doesn't exist, create it\n",
    "if(!dir.exists(opt$outdir)) {\n",
    "  print(paste('Creating output directory',opt$outdir))\n",
    "  system(paste('mkdir -p',opt$outdir))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "get.labs <- function(comps) {\n",
    "  ## Load the data, subset to shared dog-human genes\n",
    "  dat <- read.table(paste(opt$datadir, 'BRCA_rnaseq_paired_noMets.t.txt',sep='/'), sep='\\t', header=T, row.names=1, check.names=F)\n",
    "  dat <- as.matrix(dat) + 1\n",
    "  dat <- log(dat, base=2)\n",
    "  rownames(dat) <- substr(rownames(dat),1,15)\n",
    "\n",
    "  labs <- read.table(paste(opt$datadir,'BRCA_PAM50_labels.csv',sep='/'), sep=',', header=TRUE, row.names=1, check.names=FALSE, stringsAsFactors=FALSE)\n",
    "\n",
    "  ## Drop to Basal vs Luminal comparison\n",
    "  if('Luminal' %in% comps) {\n",
    "    labs[which(labs$PAM50=='LumB'),'PAM50'] <- 'Luminal'\n",
    "    labs[which(labs$PAM50=='LumA'),'PAM50'] <- 'Luminal'\n",
    "    labs <- labs[ which(labs$PAM50 %in% comps),,drop=FALSE ]\n",
    "    ids <- intersect( rownames(labs), rownames(dat) )\n",
    "    dat <- dat[ids,]\n",
    "    labs <- labs[ids,]\n",
    "  } else if('TUMOR' %in% comps) {\n",
    "    labs <- read.table(paste(opt$datadir,'BRCA_tumorVSnormal_paired.txt',sep='/'), header=F, row.names=1, check.names=F)\n",
    "    rownames(labs) <- substr(rownames(labs),1,15)\n",
    "    ids <- intersect( rownames(labs), rownames(dat) )\n",
    "    dat <- dat[ids,]\n",
    "    labs <- labs[ids,]\n",
    "  } else {\n",
    "    labs <- labs[labs[,1] %in% comps,,drop=FALSE]\n",
    "    ids <- intersect( rownames(labs), rownames(dat) )\n",
    "    dat <- dat[ids,]\n",
    "    labs <- labs[ids,]\n",
    "  }\n",
    "\n",
    "  genes.all <- read.table(paste(opt$datadir,'Canine_Human_Gene_Conversion.txt',sep='/'), sep='\\t', header=T, stringsAsFactors=F) # TODO: Update filename & location\n",
    "  dat <- dat[,colnames(dat) %in% genes.all$Hum_Symb]\n",
    "  return( list(dat, labs) )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "load.peps <- function(fn) {\n",
    "  ## Load the PEPs\n",
    "  peps <- read.table(fn, sep=',', header=TRUE, stringsAsFactors=FALSE)\n",
    "  pep.genes <- c( peps[peps$Adenoma_Expression_Pattern < 0.05,'HumanSymbol'], peps[peps$Tumor_Expression_Pattern < 0.05,'HumanSymbol'], peps[peps$Carcinoma_Expression_Pattern < 0.05,'HumanSymbol'] )\n",
    "  pep.lists <- data.frame(Gene=pep.genes, PEP=rep(NA, length(pep.genes)))\n",
    "  pep.lists[ pep.lists$Gene %in% peps[peps$Adenoma_Expression_Pattern < 0.05,'HumanSymbol'], 'PEP'] <- 'Adenoma'\n",
    "  pep.lists[ pep.lists$Gene %in% peps[peps$Carcinoma_Expression_Pattern < 0.05,'HumanSymbol'], 'PEP'] <- 'Carcinoma'\n",
    "  pep.lists[ pep.lists$Gene %in% peps[peps$Tumor_Expression_Pattern < 0.05,'HumanSymbol'], 'PEP'] <- 'Tumor'\n",
    "  peps <- pep.lists\n",
    "\n",
    "  return(peps)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code to run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Loading PEPs.\"\n"
     ]
    }
   ],
   "source": [
    "## Load the data - TODO do I need to load anything aside from the peps?\n",
    "print(\"Loading PEPs.\")\n",
    "peps <- load.peps(paste(opt$workingdir, opt$PEP, sep='/'))\n",
    "dat <- read.table(paste(opt$datadir, 'BRCA_rnaseq_paired_noMets.t.txt',sep='/'), sep='\\t', header=T, row.names=1, check.names=F)\n",
    "dat <- as.matrix(dat) + 1\n",
    "dat <- log(dat, base=2)\n",
    "rownames(dat) <- substr(rownames(dat),1,15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Loading BRCA labels and data.\"\n"
     ]
    }
   ],
   "source": [
    "## Load the labels and subset to the labeled data\n",
    "print('Loading BRCA labels and data.')\n",
    "labs <- read.table(paste(opt$datadir,'BRCA_PAM50_labels.csv',sep='/'), sep=',', header=TRUE, row.names=1, check.names=FALSE, stringsAsFactors=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Subsetting to homologous genes.\"\n"
     ]
    }
   ],
   "source": [
    "## Subset to shared samples\n",
    "ids <- intersect( rownames(labs), rownames(dat) )\n",
    "dat <- dat[ids,]\n",
    "labs <- labs[ids,,drop=FALSE]\n",
    "\n",
    "## Subset to homologous genes\n",
    "print(\"Subsetting to homologous genes.\")\n",
    "genes.all <- read.table(paste(opt$datadir,'Canine_Human_Gene_Conversion.txt',sep='/'), sep='\\t', header=T, stringsAsFactors=F) # TODO: Update filename & location\n",
    "dat <- dat[,colnames(dat) %in% genes.all$Hum_Symb]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate standard deviation and list of non-PEP genes, for later comparisons\n",
    "sd.genes <- sd(dat)\n",
    "non.pep.genes <- colnames(dat)[ !colnames(dat) %in% peps$Gene ]\n",
    "\n",
    "## Define the list of comparisons we want to make\n",
    "comparisons <- list( c('TUMOR','NORMAL'), c('Basal','Luminal'), c('LumA','Normal'), c('LumB','Normal'), c('Luminal','Normal'), c('Basal','Normal') )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## For each comparison & PEP pair, calculate PAGE score\n",
    "write('PAGE scores',file=paste(opt$outdir,'PAGE.txt', sep='/'))\n",
    "print(\"Calculating PAGE score for each PEP/comparison pair.\")\n",
    "for( comps in comparisons ) {\n",
    "\n",
    "  ## Print the comparison we're currently doing\n",
    "  write('', file=paste(opt$outdir,'PAGE.txt', sep='/'), append=TRUE)\n",
    "  write(paste(comps[1], comps[2]), file=paste(opt$outdir,'PAGE.txt', sep='/'), append=TRUE)\n",
    "  print(comps)\n",
    "\n",
    "  ## For each PEP, calculate PAGE\n",
    "  for( pep.name in c('Adenoma', 'Carcinoma', 'Tumor') ) {\n",
    "\n",
    "    # Load the applicable labels for this comparison\n",
    "    dat.list <- get.labs(comps)\n",
    "    dat <- dat.list[[1]]\n",
    "    labs <- dat.list[[2]]\n",
    "\n",
    "    ## Get the list of genes in the current PEP\n",
    "    genes <- c(non.pep.genes, as.character(peps[ peps$PEP==pep.name, 'Gene' ]))\n",
    "    genes <- genes[ genes %in% colnames(dat) ]\n",
    "\n",
    "    ## Calculate and print PAGE score for this comparison\n",
    "    res <- sapply(genes, function(gene) { ( (mean( dat[labs==comps[1],gene] ) - mean( dat[labs==comps[2],gene] ) ) * sqrt(sum(peps$PEP==pep.name)) ) / sd.genes } )\n",
    "    temp <- data.frame( PAGE=res, PEP=names(res) %in% peps[ peps$PEP==pep.name,'Gene' ] )\n",
    "    write(paste(pep.name,signif(wilcox.test( PAGE ~ PEP, data=temp )$p.value,digits=3), sep='\\t'), file=paste(opt$outdir,'PAGE.txt', sep='/'), append=TRUE)\n",
    "    print( paste(pep.name,signif(wilcox.test( PAGE ~ PEP, data=temp )$p.value,digits=3)) )\n",
    "\n",
    "  } # End PEPs loop\n",
    "} # End comps loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Results stored in ./output/PAGE.txt\"\n",
      "[1] \"Success!\"\n"
     ]
    }
   ],
   "source": [
    "print(paste('Results stored in',paste(opt$outdir,'PAGE.txt', sep='/')) )\n",
    "print('Success!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
